mmdet/models/detectors/semi_base.py

```diff
@@ -0,0 +1,297 @@
+        self.student = MODELS.build(detector)
+        self.teacher = MODELS.build(detector)
+        self.semi_train_cfg = semi_train_cfg
+        self.semi_test_cfg = semi_test_cfg
+        if self.semi_train_cfg.get('freeze_teacher', True) is True:
+            self.freeze(self.teacher)
+            self._ddp_params_and_buffers_to_ignore = [
+                'teacher.' + k for k in self.teacher.state_dict().keys()
+            ]
+
+    @staticmethod
+    def freeze(model: nn.Module):
+        """Freeze the model."""
+        model.eval()
+        for param in model.parameters():
+            param.requires_grad = False
+
+    def loss(self, multi_batch_inputs: Dict[str, Tensor],
+             multi_batch_data_samples: Dict[str, SampleList]) -> dict:
+        """Calculate losses from multi-branch inputs and data samples.
+
+        Args:
+            multi_batch_inputs (Dict[str, Tensor]): The dict of multi-branch
+                input images, each value with shape (N, C, H, W).
+                Each value should usually be mean centered and std scaled.
+            multi_batch_data_samples (Dict[str, List[:obj:`DetDataSample`]]):
+                The dict of multi-branch data samples.
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        losses = dict()
+        losses.update(**self.gt_loss(multi_batch_inputs['sup'],
+                                     multi_batch_data_samples['sup']))
+        origin_pseudo_instances, batch_info = self.get_pseudo_instances(
+            multi_batch_inputs['unsup_teacher'],
+            multi_batch_data_samples['unsup_teacher'])
+        multi_batch_data_samples[
+            'unsup_student'] = self.project_pseudo_instances(
+                origin_pseudo_instances,
+                multi_batch_data_samples['unsup_student'])
+        losses.update(**self.pseudo_loss(
+            multi_batch_inputs['unsup_student'],
+            multi_batch_data_samples['unsup_student'], batch_info))
+        return losses
+
+    def gt_loss(self, batch_inputs: Tensor,
+                batch_data_samples: SampleList) -> dict:
+        """Calculate losses from a batch of inputs and ground-truth data
+        samples.
+
+        Args:
+            batch_inputs (Tensor): Input images of shape (N, C, H, W).
+                These should usually be mean centered and std scaled.
+            batch_data_samples (List[:obj:`DetDataSample`]): The batch
+                data samples. It usually includes information such
+                as `gt_instance` or `gt_panoptic_seg` or `gt_sem_seg`.
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        gt_loss = {
+            'sup_' + k: v
+            for k, v in self.weight(
+                self.student.loss(batch_inputs, batch_data_samples),
+                self.semi_train_cfg.get('sup_weight', 1.)).items()
+        }
+        return gt_loss
+
+    def pseudo_loss(self,
```

Comment:
pseudo_supervised_loss

Reply:
loss_by_pseudo_instance ?

Reply:
loss_by_pseudo_label might be more clear